src/audio_core/opus/decoder.cpp

```diff
@@ -0,0 +1,183 @@
+// SPDX-FileCopyrightText: Copyright 2023 yuzu Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "audio_core/opus/decoder.h"
+#include "audio_core/opus/hardware_opus.h"
+#include "audio_core/opus/parameters.h"
+#include "common/alignment.h"
+#include "common/swap.h"
+#include "core/core.h"
+
+namespace AudioCore::OpusDecoder {
+namespace {
+OpusPacketHeader ReverseHeader(OpusPacketHeader header) {
+    OpusPacketHeader out;
+    out.size = Common::swap32(header.size);
+    out.final_range = Common::swap32(header.final_range);
+    return out;
+}
+} // namespace
+
+OpusDecoder::OpusDecoder(Core::System& system_, HardwareOpus& hardware_opus_)
+    : system{system_}, hardware_opus{hardware_opus_} {}
+
+OpusDecoder::~OpusDecoder() {
+    if (decode_object_initialized) {
+        hardware_opus.ShutdownDecodeObject(shared_buffer.get(), shared_buffer_size);
+    }
+}
+
+Result OpusDecoder::Initialize(OpusParametersEx& params, Kernel::KTransferMemory* transfer_memory,
+                               u64 transfer_memory_size) {
+    auto frame_size{params.use_large_frame_size ? 5760 : 1920};
+    shared_buffer_size = transfer_memory_size;
+    shared_buffer = std::make_unique<u8[]>(shared_buffer_size);
+    shared_memory_mapped = true;
+
+    buffer_size =
+        Common::AlignUp((frame_size * params.channel_count) / (48'000 / params.sample_rate), 16);
+
+    out_data = {shared_buffer.get() + shared_buffer_size - buffer_size, buffer_size};
+    size_t in_data_size{0x600u};
+    in_data = {out_data.data() - in_data_size, in_data_size};
+
+    if (hardware_opus.InitializeDecodeObject(params.sample_rate, params.channel_count,
+                                             shared_buffer.get(),
+                                             shared_buffer_size) != ResultSuccess) {
+        if (shared_memory_mapped) {
+            shared_memory_mapped = false;
+            R_TRY(hardware_opus.UnmapMemory(shared_buffer.get(), shared_buffer_size));
+        }
+    }

```

Comment:
I would expect that the decode object initialization is transactional, and doesn't leave a lingering mapping if initialization fails. If that's true, then
```suggestion
    R_TRY(hardware_opus.InitializeDecodeObject(params.sample_rate, params.channel_count, shared_buffer.get(), shared_buffer_size));
```

Reply:
I don't know what you mean by this. It calls unmap if init fails.

Reply:
In that case, move the unmap code into the scope guard from the above comment that marks the shared memory as unmapped. Then this becomes R_TRY.